src/librustc_mir/interpret/memory.rs

```diff
@@ -297,17 +297,21 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> Memory<'a, 'mir, 'tcx, M> {
     /// Check if the pointer is "in-bounds". Notice that a pointer pointing at the end
     /// of an allocation (i.e., at the first *inaccessible* location) *is* considered
     /// in-bounds!  This follows C's/LLVM's rules.
-    /// This function also works for deallocated allocations.
-    /// Use `.get(ptr.alloc_id)?.check_bounds_ptr(ptr)` if you want to force the allocation
-    /// to still be live.
     /// If you want to check bounds before doing a memory access, better first obtain
     /// an `Allocation` and call `check_bounds`.
-    pub fn check_bounds_ptr_maybe_dead(
+    pub fn check_bounds_ptr(
         &self,
         ptr: Pointer<M::PointerTag>,
+        liveness: InboundsCheck,
     ) -> EvalResult<'tcx, Align> {
-        let (allocation_size, align) = self.get_size_and_align(ptr.alloc_id);
-        ptr.check_in_alloc(allocation_size, InboundsCheck::MaybeDead)?;
+        let (allocation_size, align) = match liveness {
+            InboundsCheck::Live => {
+                let alloc = self.get(ptr.alloc_id)?;
+                (Size::from_bytes(alloc.bytes.len() as u64), alloc.align)
+            }
+            InboundsCheck::MaybeDead => self.get_size_and_align(ptr.alloc_id),
+        };
+        ptr.check_in_alloc(allocation_size, liveness)?;
```

Comment:
Is `check_in_alloc` still used anywhere else?  Might be worth to inline it here, because it clearly has a confusing API.